File: data/services/mill-data-http-service/src/main/java/io/qpointz/mill/data/backend/access/http/explorer/SchemaExplorerController.java
package io.qpointz.mill.data.backend.access.http.explorer;

import io.qpointz.mill.data.backend.access.http.explorer.dto.SearchResultDto;
import io.qpointz.mill.data.backend.access.http.explorer.dto.TreeNodeDto;
import io.qpointz.mill.metadata.domain.MetadataEntity;
import io.qpointz.mill.metadata.domain.MetadataType;
import io.qpointz.mill.metadata.service.MetadataService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@RestController
@RequestMapping("/api/metadata/v1/explorer")
@ConditionalOnBean(MetadataService.class)
@RequiredArgsConstructor
@CrossOrigin(origins = {"http://localhost:5173", "http://localhost:8080"})
public class SchemaExplorerController {

    private final MetadataService metadataService;

    @GetMapping("/tree")
    public ResponseEntity<List<TreeNodeDto>> getTree(
            @RequestParam(name = "schema", required = false) String schema,
            @RequestParam(name = "scope", required = false, defaultValue = "global") String scope) {
        List<MetadataEntity> entities = metadataService.findAll();
        List<MetadataEntity> filtered = entities.stream()
                .filter(e -> schema == null || Objects.equals(e.getSchemaName(), schema))
                .toList();
        return ResponseEntity.ok(buildTree(filtered, scope));
    }

    @GetMapping("/search")
    public ResponseEntity<List<SearchResultDto>> search(
            @RequestParam(name = "q") String q,
            @RequestParam(name = "type", required = false) String type,
            @RequestParam(name = "scope", required = false, defaultValue = "global") String scope) {
        String query = q.toLowerCase();
        List<SearchResultDto> results = metadataService.findAll().stream()
                .filter(e -> type == null || e.getType().name().equalsIgnoreCase(type))
                .filter(e -> matchesQuery(e, query))
                .map(e -> toSearchResult(e, scope))
                .sorted(Comparator.comparing(SearchResultDto::getName))
                .toList();
        return ResponseEntity.ok(results);
    }

    @GetMapping("/lineage")
    public ResponseEntity<Map<String, Object>> getLineage(
            @RequestParam(name = "table") String table,
            @RequestParam(name = "depth", required = false, defaultValue = "1") int depth) {
        Map<String, Object> lineage = Map.of(
                "table", table,
                "depth", depth,
                "upstream", List.of(),
                "downstream", List.of());
        return ResponseEntity.ok(lineage);
    }

    private List<TreeNodeDto> buildTree(List<MetadataEntity> entities, String scope) {
        Map<String, List<MetadataEntity>> bySchema = entities.stream()
                .filter(e -> e.getSchemaName() != null)
                .collect(Collectors.groupingBy(MetadataEntity::getSchemaName));

        return bySchema.entrySet().stream()
                .map(entry -> {
                    String schemaName = entry.getKey();
                    List<MetadataEntity> schemaEntities = entry.getValue();

                    List<MetadataEntity> tableEntities = schemaEntities.stream()
                            .filter(e -> e.getType() == MetadataType.TABLE)
                            .sorted(Comparator.comparing(MetadataEntity::getTableName))
                            .toList();

                    List<TreeNodeDto> tables = tableEntities.stream()
                            .map(tableEntity -> {
                                List<TreeNodeDto> attributeNodes = schemaEntities.stream()
                                        .filter(e -> e.getType() == MetadataType.ATTRIBUTE)
                                        .filter(e -> Objects.equals(e.getTableName(), tableEntity.getTableName()))
                                        .sorted(Comparator.comparing(MetadataEntity::getAttributeName))
                                        .map(attr -> toTreeNode(attr, scope, false))
                                        .toList();

                                TreeNodeDto tableNode = toTreeNode(tableEntity, scope, true);
                                tableNode.setChildren(attributeNodes);
                                tableNode.setHasChildren(!attributeNodes.isEmpty());
                                return tableNode;
                            })
                            .sorted(Comparator.comparing(TreeNodeDto::getName))
                            .toList();

                    return TreeNodeDto.builder()
                            .id(schemaName)
                            .name(schemaName)
                            .type(MetadataType.SCHEMA)
                            .displayName(schemaName)
                            .children(tables)
                            .hasChildren(!tables.isEmpty())
                            .build();
                })
                .sorted(Comparator.comparing(TreeNodeDto::getName))
                .toList();
    }

    @SuppressWarnings("unchecked")
    private TreeNodeDto toTreeNode(MetadataEntity entity, String scope, boolean includeChildren) {
        String displayName = entity.getFacet("descriptive", scope, Map.class)
                .map(f -> (String) ((Map<?, ?>) f).get("displayName"))
                .orElse(getEntityName(entity));
        String description = entity.getFacet("descriptive", scope, Map.class)
                .map(f -> (String) ((Map<?, ?>) f).get("description"))
                .orElse(null);

        return TreeNodeDto.builder()
                .id(entity.getId())
                .name(getEntityName(entity))
                .type(entity.getType())
                .displayName(displayName)
                .description(description)
                .children(includeChildren ? List.of() : null)
                .hasChildren(false)
                .build();
    }

    private String getEntityName(MetadataEntity entity) {
        if (entity.getAttributeName() != null) return entity.getAttributeName();
        if (entity.getTableName() != null) return entity.getTableName();
        if (entity.getSchemaName() != null) return entity.getSchemaName();
        return entity.getId();
    }

    @SuppressWarnings("unchecked")
    private boolean matchesQuery(MetadataEntity entity, String query) {
        if (entity.getId() != null && entity.getId().toLowerCase().contains(query)) return true;
        if (entity.getTableName() != null && entity.getTableName().toLowerCase().contains(query)) return true;
        if (entity.getAttributeName() != null && entity.getAttributeName().toLowerCase().contains(query)) return true;
        Optional<Map> descriptive = entity.getFacet("descriptive", "global", Map.class);
        if (descriptive.isPresent()) {
            Map<?, ?> desc = descriptive.get();
            if (desc.get("displayName") != null && desc.get("displayName").toString().toLowerCase().contains(query))
                return true;
            if (desc.get("description") != null && desc.get("description").toString().toLowerCase().contains(query))
                return true;
        }
        return false;
    }

    private SearchResultDto toSearchResult(MetadataEntity entity, String scope) {
        @SuppressWarnings("unchecked")
        String displayName = entity.getFacet("descriptive", scope, Map.class)
                .map(f -> (String) ((Map<?, ?>) f).get("displayName"))
                .orElse(entity.getTableName() != null ? entity.getTableName() : entity.getId());
        @SuppressWarnings("unchecked")
        String description = entity.getFacet("descriptive", scope, Map.class)
                .map(f -> (String) ((Map<?, ?>) f).get("description"))
                .orElse(null);

        return SearchResultDto.builder()
                .id(entity.getId())
                .name(entity.getTableName() != null ? entity.getTableName() : entity.getId())
                .type(entity.getType())
                .displayName(displayName)
                .description(description)
                .location(buildLocation(entity))
                .build();
    }

    private String buildLocation(MetadataEntity entity) {
        if (entity.getAttributeName() != null) {
            return String.format("%s.%s.%s", entity.getSchemaName(), entity.getTableName(), entity.getAttributeName());
        }
        if (entity.getTableName() != null) {
            return String.format("%s.%s", entity.getSchemaName(), entity.getTableName());
        }
        if (entity.getSchemaName() != null) return entity.getSchemaName();
        return entity.getId();
    }
}

File: metadata/mill-metadata-service/src/main/kotlin/io/qpointz/mill/metadata/api/SchemaExplorerController.kt
package io.qpointz.mill.metadata.api

import io.qpointz.mill.metadata.api.dto.SearchResultDto
import io.qpointz.mill.metadata.api.dto.TreeNodeDto
import io.qpointz.mill.metadata.domain.MetadataEntity
import io.qpointz.mill.metadata.domain.MetadataType
import io.qpointz.mill.metadata.service.MetadataService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/** Endpoints for schema navigation and metadata discovery. */
@RestController
@RequestMapping("/api/metadata/v1/explorer")
@CrossOrigin(origins = ["http://localhost:5173", "http://localhost:8080"])
class SchemaExplorerController(private val metadataService: MetadataService) {

    @GetMapping("/tree")
    fun getTree(
        @RequestParam(name = "schema", required = false) schema: String?,
        @RequestParam(name = "scope", required = false, defaultValue = "global") scope: String
    ): ResponseEntity<List<TreeNodeDto>> {
        val filtered = metadataService.findAll()
            .filter { schema == null || it.schemaName == schema }
        return ResponseEntity.ok(buildTree(filtered, scope))
    }

    @GetMapping("/search")
    fun search(
        @RequestParam(name = "q") q: String,
        @RequestParam(name = "type", required = false) type: String?,
        @RequestParam(name = "scope", required = false, defaultValue = "global") scope: String
    ): ResponseEntity<List<SearchResultDto>> {
        val query = q.lowercase()
        val results = metadataService.findAll().asSequence()
            .filter { type == null || it.type?.name?.equals(type, ignoreCase = true) == true }
            .filter { matchesQuery(it, query) }
            .map { toSearchResult(it, scope) }
            .sortedWith(compareBy<SearchResultDto> { it.name ?: "" })
            .toList()
        return ResponseEntity.ok(results)
    }

    @GetMapping("/lineage")
    fun getLineage(
        @RequestParam(name = "table") table: String,
        @RequestParam(name = "depth", required = false, defaultValue = "1") depth: Int
    ): ResponseEntity<Map<String, Any>> {
        val lineage = mapOf(
            "table" to table,
            "depth" to depth,
            "upstream" to emptyList<String>(),
            "downstream" to emptyList<String>()
        )
        return ResponseEntity.ok(lineage)
    }

    private fun buildTree(entities: List<MetadataEntity>, scope: String): List<TreeNodeDto> {
        val bySchema = entities
            .filter { it.schemaName != null }
            .groupBy { it.schemaName!! }

        return bySchema.entries.asSequence()
            .map { (schemaName, schemaEntities) ->
                val tableEntities = schemaEntities.asSequence()
                    .filter { it.type == MetadataType.TABLE }
                    .sortedBy { it.tableName }
                    .toList()

                val tables = tableEntities.asSequence()
                    .map { tableEntity ->
                        val attributeNodes = schemaEntities.asSequence()
                            .filter { it.type == MetadataType.ATTRIBUTE }
                            .filter { it.tableName == tableEntity.tableName }
                            .sortedBy { it.attributeName }
                            .map { toTreeNode(it, scope, includeChildren = false) }
                            .toList()

                        toTreeNode(tableEntity, scope, includeChildren = true).apply {
                            children = attributeNodes
                            hasChildren = attributeNodes.isNotEmpty()
                        }
                    }
                    .sortedWith(compareBy<TreeNodeDto> { it.name ?: "" })
                    .toList()

                TreeNodeDto(
                    id = schemaName,
                    name = schemaName,
                    type = MetadataType.SCHEMA,
                    displayName = schemaName,
                    children = tables,
                    hasChildren = tables.isNotEmpty()
                )
            }
            .sortedWith(compareBy<TreeNodeDto> { it.name ?: "" })
            .toList()
    }

    @Suppress("UNCHECKED_CAST")
    private fun toTreeNode(entity: MetadataEntity, scope: String, includeChildren: Boolean): TreeNodeDto {
        val descriptive = entity.getFacet("descriptive", scope, Map::class.java)
            .map { it as Map<*, *> }
        val displayName = descriptive.map { it["displayName"] as? String }.orElse(getEntityName(entity))
        val description = descriptive.map { it["description"] as? String }.orElse(null)

        return TreeNodeDto(
            id = entity.id,
            name = getEntityName(entity),
            type = entity.type,
            displayName = displayName,
            description = description,
            children = if (includeChildren) emptyList() else null,
            hasChildren = false
        )
    }

    private fun getEntityName(entity: MetadataEntity): String =
        entity.attributeName
            ?: entity.tableName
            ?: entity.schemaName
            ?: entity.id
            ?: ""

    @Suppress("UNCHECKED_CAST")
    private fun matchesQuery(entity: MetadataEntity, query: String): Boolean {
        if (entity.id?.lowercase()?.contains(query) == true) return true
        if (entity.tableName?.lowercase()?.contains(query) == true) return true
        if (entity.attributeName?.lowercase()?.contains(query) == true) return true

        val descriptive = entity.getFacet("descriptive", "global", Map::class.java)
            .map { it as Map<*, *> }
        if (descriptive.isPresent) {
            val map = descriptive.get()
            if (map["displayName"]?.toString()?.lowercase()?.contains(query) == true) return true
            if (map["description"]?.toString()?.lowercase()?.contains(query) == true) return true
        }
        return false
    }

    @Suppress("UNCHECKED_CAST")
    private fun toSearchResult(entity: MetadataEntity, scope: String): SearchResultDto {
        val descriptive = entity.getFacet("descriptive", scope, Map::class.java)
            .map { it as Map<*, *> }
        val displayName = descriptive.map { it["displayName"] as? String }
            .orElse(entity.tableName ?: entity.id)
        val description = descriptive.map { it["description"] as? String }.orElse(null)

        return SearchResultDto(
            id = entity.id,
            name = entity.tableName ?: entity.id,
            type = entity.type,
            displayName = displayName,
            description = description,
            location = buildLocation(entity)
        )
    }

    private fun buildLocation(entity: MetadataEntity): String {
        if (entity.attributeName != null) {
            return "${entity.schemaName}.${entity.tableName}.${entity.attributeName}"
        }
        if (entity.tableName != null) {
            return "${entity.schemaName}.${entity.tableName}"
        }
        return entity.schemaName ?: entity.id ?: ""
    }
}
